File: backend/src/app/schemas/escalation.py
"""Pydantic schemas for Escalation model"""
from datetime import datetime
from uuid import UUID
from pydantic import BaseModel, Field
from typing import Optional
from enum import Enum as PyEnum


class EscalationStatus(str, PyEnum):
    """Status of an escalation"""
    PENDING = "pending"
    ASSIGNED = "assigned"
    IN_PROGRESS = "in_progress"
    RESOLVED = "resolved"
    FALSE_ALARM = "false_alarm"
    CANCELLED = "cancelled"


class EscalationPriority(str, PyEnum):
    """Priority level of escalation"""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class EscalationBase(BaseModel):
    """Base escalation schema"""
    title: str = Field(..., min_length=1, max_length=255)
    description: str = Field(..., min_length=1)
    priority: EscalationPriority = EscalationPriority.MEDIUM
    zone_uuid: UUID
    camera_uuid: Optional[UUID] = None


class EscalationCreate(EscalationBase):
    """Schema for creating escalation"""
    pass


class EscalationUpdate(BaseModel):
    """Schema for updating escalation"""
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, min_length=1)
    priority: Optional[EscalationPriority] = None
    assigned_to_uuid: Optional[UUID] = None
    status: Optional[EscalationStatus] = None


class EscalationAction(BaseModel):
    """Schema for taking action on escalation"""
    action_taken: str = Field(..., min_length=1)


class EscalationResolve(BaseModel):
    """Schema for resolving escalation"""
    pass


class EscalationFalseAlarm(BaseModel):
    """Schema for marking as false alarm"""
    pass


class UserMinimal(BaseModel):
    """Minimal user information"""
    uuid: UUID
    name: Optional[str] = None
    email: str
    role: str

    class Config:
        from_attributes = True


class EscalationResponse(EscalationBase):
    """Response schema for escalation"""
    uuid: UUID
    assigned_to_uuid: Optional[UUID] = None
    created_by_uuid: UUID
    zone_name: Optional[str] = None
    camera_name: Optional[str] = None
    assigned_to_name: Optional[str] = None
    created_by_name: Optional[str] = None
    status: EscalationStatus
    priority: EscalationPriority
    action_taken: Optional[str] = None
    is_acted_upon: bool
    is_false_alarm: bool
    acted_at: Optional[datetime] = None
    created_at: datetime
    updated_at: Optional[datetime] = None
    resolved_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class EscalationList(BaseModel):
    """List response for escalations"""
    uuid: UUID
    title: str
    priority: EscalationPriority
    status: EscalationStatus
    zone_uuid: UUID
    assigned_to_uuid: Optional[UUID] = None
    created_by_uuid: UUID
    is_acted_upon: bool
    is_false_alarm: bool
    created_at: datetime
    resolved_at: Optional[datetime] = None
    zone_name: Optional[str] = None
    camera_name: Optional[str] = None
    assigned_to_name: Optional[str] = None
    created_by_name: Optional[str] = None

    # Minimal user info
    assigned_to: Optional[UserMinimal] = None
    created_by: Optional[UserMinimal] = None

    class Config:
        from_attributes = True

File: frontend/src/services/escalation.ts
/**
 * Escalation API Service
 * Handles all escalation-related API calls
 */

const BASE_URL = `${import.meta.env.VITE_API_URL || 'http://localhost:8000'}/api/v1`;

const getAuthToken = () => {
  const localToken = localStorage.getItem('authToken');
  if (localToken) return localToken;

  const rawUser = localStorage.getItem('currentUser');
  if (!rawUser) return null;

  try {
    const parsed = JSON.parse(rawUser) as { token?: string };
    return parsed.token || null;
  } catch {
    return null;
  }
};

const getAuthHeader = (): HeadersInit => {
  const token = getAuthToken();
  if (!token) return {};
  return { Authorization: `Bearer ${token}` };
};

export interface Escalation {
  uuid: string;
  zone_uuid: string;
  zone_name?: string;
  title: string;
  description: string;
  priority: "low" | "medium" | "high" | "critical";
  status: "pending" | "assigned" | "in_progress" | "resolved" | "false_alarm" | "cancelled";
  assigned_to?: string;
  assigned_to_uuid: string | null;
  assigned_to_name?: string;
  created_by_uuid: string;
  action_taken: string | null;
  is_acted_upon: boolean;
  is_false_alarm: boolean;
  acted_at: string | null;
  camera_uuid: string | null;
  camera_name?: string;
  created_at: string;
  updated_at: string | null;
  resolved_at: string | null;
}

export interface EscalationCreate {
  zone_uuid: string;
  title: string;
  description: string;
  priority: "low" | "medium" | "high" | "critical";
  camera_uuid?: string;
}

export interface EscalationUpdate {
  title?: string;
  description?: string;
  priority?: "low" | "medium" | "high" | "critical";
  assigned_to_uuid?: string;
  status?: string;
}

export interface EscalationAction {
  action_taken: string;
}

export interface EscalationStats {
  total: number;
  pending: number;
  assigned: number;
  in_progress: number;
  resolved: number;
  false_alarms: number;
  critical: number;
}

class EscalationService {
  /**
   * Create a new escalation
   */
  async createEscalation(data: EscalationCreate): Promise<Escalation> {
    const response = await fetch(`${BASE_URL}/escalations`, {
      method: "POST",
      headers: { "Content-Type": "application/json", ...getAuthHeader() },
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      throw new Error(`Failed to create escalation: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get list of escalations with optional filtering
   */
  async listEscalations(params?: {
    zone_uuid?: string;
    status_filter?: string;
    priority_filter?: string;
  }): Promise<Escalation[]> {
    const queryParams = new URLSearchParams();
    if (params?.zone_uuid) queryParams.append("zone_uuid", params.zone_uuid);
    if (params?.status_filter) queryParams.append("status_filter", params.status_filter);
    if (params?.priority_filter) queryParams.append("priority_filter", params.priority_filter);

    const url = queryParams.toString()
      ? `${BASE_URL}/escalations?${queryParams.toString()}`
      : `${BASE_URL}/escalations`;

    const response = await fetch(url, {
      headers: { ...getAuthHeader() },
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch escalations: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get escalation details
   */
  async getEscalation(uuid: string): Promise<Escalation> {
    const response = await fetch(`${BASE_URL}/escalations/${uuid}`, {
      headers: { ...getAuthHeader() },
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch escalation: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Update an escalation (admin only)
   */
  async updateEscalation(uuid: string, data: EscalationUpdate): Promise<Escalation> {
    const response = await fetch(`${BASE_URL}/escalations/${uuid}`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json", ...getAuthHeader() },
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      throw new Error(`Failed to update escalation: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Act on an escalation (mark as acted upon)
   */
  async actOnEscalation(uuid: string, actionTaken: string): Promise<Escalation> {
    const response = await fetch(`${BASE_URL}/escalations/${uuid}/act`, {
      method: "POST",
      headers: { "Content-Type": "application/json", ...getAuthHeader() },
      body: JSON.stringify({ action_taken: actionTaken }),
    });

    if (!response.ok) {
      throw new Error(`Failed to act on escalation: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Resolve an escalation
   */
  async resolveEscalation(uuid: string): Promise<Escalation> {
    const response = await fetch(`${BASE_URL}/escalations/${uuid}/resolve`, {
      method: "POST",
      headers: { ...getAuthHeader() },
    });

    if (!response.ok) {
      throw new Error(`Failed to resolve escalation: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Mark escalation as false alarm
   */
  async markFalseAlarm(uuid: string): Promise<Escalation> {
    const response = await fetch(`${BASE_URL}/escalations/${uuid}/false-alarm`, {
      method: "POST",
      headers: { ...getAuthHeader() },
    });

    if (!response.ok) {
      throw new Error(`Failed to mark as false alarm: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get escalation statistics
   */
  async getStats(): Promise<EscalationStats> {
    const response = await fetch(`${BASE_URL}/escalations/stats/summary`, {
      headers: { ...getAuthHeader() },
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch stats: ${response.statusText}`);
    }

    return response.json();
  }
}

export const escalationService = new EscalationService();
